test/test_cuda.py

```diff
@@ -1901,6 +1901,237 @@ def worker(rank):
+        self.assertTrue(new_scale.item(), 2.0)
+
+    def _create_scaling_models_optimizers(self):
+        # Create a module+optimizer that will use scaling, and a control module+optimizer
+        # that will not use scaling, against which the scaling-enabled module+optimizer can be compared.
+        mod_control = torch.nn.Sequential(torch.nn.Linear(8, 8), torch.nn.Linear(8, 8)).cuda()
+        mod_scaling = torch.nn.Sequential(torch.nn.Linear(8, 8), torch.nn.Linear(8, 8)).cuda()
+        for c, s in zip(mod_control.parameters(), mod_scaling.parameters()):
+            s.data.copy_(c.data)
+
+        opt_control = torch.optim.SGD(mod_control.parameters(), lr=1.0)
+        opt_scaling = torch.optim.SGD(mod_scaling.parameters(), lr=1.0)
+
+        return mod_control, mod_scaling, opt_control, opt_scaling
+
+    def _create_scaling_case(self):
+        data = [(torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda")),
+                (torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda")),
+                (torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda")),
+                (torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda"))]
+
+        loss_fn = torch.nn.MSELoss().cuda()
+
+        skip_iter = 3
+
+        return self._create_scaling_models_optimizers() + (data, loss_fn, skip_iter)
+
+    # Genericize a bit of the testing logic to avoid too much copy-pasting below
+    def _run_scaling_case(self, run, unskipped, skipped):
+        # Ensure scaling can be disabled without changing user control flow.
+        for enabled in True, False:
+            mod_control, mod_scaling, opt_control, opt_scaling, data, loss_fn, skip_iter = self._create_scaling_case()
+
+            # For functionality, test with a modest initial scale, and an unrealistically-large growth factor
+            # so any potential errors with the growth factor handling will be magnified.
+            scaler = torch.cuda.amp.AmpScaler(init_scale=128., growth_factor=2.0, enabled=enabled)
+
+            run(data, mod_control, opt_control, scaler, loss_fn, skip_iter, False)
+            run(data, mod_scaling, opt_scaling, scaler, loss_fn, skip_iter, True)
+
+            # If scaling was enabled, the scale factor should have been multiplied by the growth factor
+            # len(data) - skipped times and the backoff factor "skipped" times.
+            if enabled:
+                net_growth = scaler.get_growth_factor()**unskipped if unskipped > 0 else 1.0
+                net_backoff = scaler.get_backoff_factor()**skipped if skipped > 0 else 1.0
+                self.assertTrue(scaler.get_scale() == (128. * net_growth * net_backoff))
+            else:
+                self.assertTrue(scaler.get_scale() == 1.0)
+
+            for c, s in zip(mod_control.parameters(), mod_scaling.parameters()):
+                self.assertTrue(torch.allclose(c, s, atol=1e-7))
+
+    def test_grad_scaling_clipping(self):
```

Comment:
I think you can remove this test case?

Reply:
I prefer to keep this one too because I already removed the "typical use" case, so this is now the case that's most similar to "typical use" for 1 optimizer without separate unscaling.